### PR TITLE
[MBL-17706][Student] Offline Studio videos

### PR DESCRIPTION
**Pull request by tamaskozmer**
### How the download works?

- Before course sync we authenticate for the Studio API:
  - We fetch LTIs for the account and search for Studio.
  - Launch Studio LTI in a headless webview.
  - Extract the user id and token with javascript.
  - Create an API entity for studio API access.
- We fetch the studio metadata for the courses that are synced. This happens before the course sync so the HTMLParser can create the correct HTML with the captions.
- Course sync takes place, we parse the HTMLs and extract what studio videos should we actually download
- Studio sync starts after the course sync. We already know what videos we need so we clean up the videos that are not needed anymore and download the videos that are needed.
- After a video download we generate the preview image and the captions and save them next to the video.

### Notes

- The authentication method is a temporary one until we receive a dedicated API for it around Q4.
- OfflineSyncWorker now works as a foreground service and will have an ongoing notification when syncing. This is needed to authenticate Studio in the background using a headless WebView, but it's a good practice anyway.
- Studio videos are course independent and will have a new item on the progress screen under the last course item.

Test plan:

- Embed a studio video with subtitles inside a page, assignment, etc.
- Sync the course for offline availability.
- Go offline and enter the page, assignment, etc.
- Check if studio videos can be played with subtitles.
- Test the offline progress screen as well, it should add the Studio progress to the total progress as well and should show a new item for Studio videos under the course items.
- Smoke test the offline sync.
- Check app update against a master build, because there were changes in the DB.

refs: MBL-17706
affects: Student
release note: Studio videos are now downloaded during an offline sync.

## Checklist

- [x] Run E2E test suite
- [x] Tested in dark mode
- [x] Tested in light mode
- [x] Approve from product

### CODE REVIEW

**Comment by inst-danger** [comment]
<p>Parent Install Page</p>

**Comment by inst-danger** [comment]
<p>Parent Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Review by kdeakinstructure — APPROVED** [review]
QA 👍

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Parent Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Parent Install Page</p>

File: libs/canvas-api-2/src/main/java/com/instructure/canvasapi2/apis/StudioApi.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import com.instructure.canvasapi2.models.StudioMediaMetadata
+import com.instructure.canvasapi2.utils.DataResult
+import retrofit2.http.GET
+import retrofit2.http.Tag
+import retrofit2.http.Url
+
+interface StudioApi {
+
+    @GET
+    suspend fun getStudioMediaMetadata(@Url url: String, @Tag params: RestParams): DataResult<List<StudioMediaMetadata>>
```
Comment (kristofnemere) [inline libs/canvas-api-2/src/main/java/com/instructure/canvasapi2/apis/StudioApi.kt:28]: I would have added the bearer as a param here, so we don't have to modify RestParams

**Review by kristofnemere — APPROVED** [review]

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Parent Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Review by hermannakos — APPROVED** [review]

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Parent Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Pull request merged**